src/main/kotlin/org/rust/ide/intentions/ExtractInlineModule.kt

```diff
@@ -0,0 +1,56 @@
+package org.rust.ide.intentions
+
+import com.intellij.codeInsight.actions.ReformatCodeProcessor
+import com.intellij.codeInsight.intention.IntentionAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.io.FileUtil
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiFileFactory
+import org.rust.lang.RustLanguage
+import org.rust.lang.core.psi.RustModDeclItem
+import org.rust.lang.core.psi.RustModItem
+import org.rust.lang.core.psi.util.childOfType
+import org.rust.lang.core.psi.util.parentOfType
+
+class ExtractInlineModule : IntentionAction {
+    override fun getFamilyName() = "Extract inline module structure"
+    override fun getText() = "Extract Inline Module"
+    override fun startInWriteAction() = true
+
+    fun createFileFromText(project: Project, text: String) =
+        PsiFileFactory.getInstance(project).createFileFromText("DUMMY.rs", RustLanguage, text)
+
+    fun createModDeclItem(project: Project, modName: String): RustModDeclItem? {
+        val file = createFileFromText(project, "mod $modName;")
+        return file?.childOfType<RustModDeclItem>()
+    }
+
+    override fun invoke(project: Project, editor: Editor?, file: PsiFile?) {
+        file ?: return
+        val offset = editor?.caretModel?.offset ?: return
+        val element = file.findElementAt(offset)?.parentOfType<RustModItem>() ?: return
+        val elementName = element.name ?: return
+        val fileName = "$elementName.rs"
+        val modFile = file.parent?.createFile(fileName)
+
+        val copyElement = element.copy() as? RustModItem ?: return
+        copyElement.mod.delete()
+        copyElement.rbrace?.delete()
+        copyElement.lbrace?.delete()
+        copyElement.nameElement?.delete()
+
+        modFile?.add(copyElement)
+        ReformatCodeProcessor(project, modFile, null, false).run()
+
+        val decl = createModDeclItem(project, elementName) ?: return
+        element.replace(decl)
+    }
+
+    override fun isAvailable(project: Project, editor: Editor?, file: PsiFile?): Boolean {
+        file ?: return false
+        val offset = editor?.caretModel?.offset ?: return false
+        var element = file.findElementAt(offset)?.parentOfType<RustModItem>() ?: return false
+        return FileUtil.getNameWithoutExtension(file.name) != (element).name
```

Comment:
I think this check is better written as `!mod.isTopLevelInFile`. That is, rename `element` -> `mod`, and create an extension method for `RustModItem`. And the better way to actually check the condition is something like `mod.parent? is RustFile`